Engine/src/Engine/ResourceManager.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "Core.h"
+#include "json.h"
+#include "Log.h"
+#include <iostream>
+#include <fstream>
+#include<sstream>
+#include <string>
+#include <filesystem>
+#include <map>
+
+//Written by Kevin Vilanova, KevinAlexV.
+//Class is intended to import and manage assets and resources used by the game engine.
+namespace Engine
+{
+
+	class ENGINE_API ResourceManager
+	{
+		
+		public:
+
+			#pragma region Singleton Instance Management
+
+			//Singletons should not be cloneable, this is to prevent clones.
+			ResourceManager(ResourceManager &other) = delete;
+			
+			//Singletons should not be assignable, this is to prevent that.
+			void operator=(const ResourceManager &) = delete;
+			
+			//This retrieves a pointer to the current instance of ResourceManager. If it doesn't exist, then one will be created and returned.
+			static ResourceManager* getInstance();
+			
+			#pragma endregion
+
+			#pragma region Set Functions
+
+			//Function to save all file paths into m_filePaths within the provided path, including subdirectories.
+			void saveFilePaths(const std::string& path);
+			#pragma endregion
+
+			#pragma region Get Functions
+			
+			//Function to return a json from the hashmap based on a provided name. Returns a nullptr if no json is found.
+			nlohmann::json getJsonData(const std::string& Name);
+
+			//Function to return a shader from the hashmap based on a provided name. Returns an empty string if no shader is found.
+			std::string getShaderData(const std::string& Name);
+			#pragma endregion
+
+		
+		protected:
+			#pragma region Constructors
+			
+						//Constructor for ResourceManager Class. Finds all files under the 'data' path and stores the paths in a map. 
+						ResourceManager();
+			#pragma endregion
+		
+		private:
+
+			
+
+			//The static instance of ResourceManager that will be returned if the resource manager is requested.
+			static ResourceManager* m_pinstance;
+			
+			//static lock m_mutex is the lock to be used for static functions. Otherwise, m_functionlock should be utilized.
+			static std::mutex m_mutex;
+			std::mutex m_functionLock;
+			
+			#pragma region File Storage Variables
+			
+			//Extensions for all files handled through resource manager.
+			std::string m_jsonFileExt = "json";
+
+			std::vector<std::string> m_shaderFileExts = { "vs", "fs" };
+			#pragma endregion
+
+			#pragma region File Storage Variables
```

Comment:
I'm going to assume these are indented correctly but are misrepresented in github ?

Reply:
Oh, is there a standard for pragma regions? I've never used them, so I'm not sure what the standard is.

Reply:
No I just mean the indenting in the entire file. Some of it seems really pushed forward, but I saw that in other reviews too so I'm assuming it's a GitHub thing

Reply:
Gotcha. Not sure, but they look fine on my end